juju/model/_idle.py

```diff
@@ -0,0 +1,215 @@
+            continue
+
+        expected_idle_since = now - idle_period
+
+        # FIXME there's some confusion about what a "busy" unit is
+        # are we ready when over the last idle_period, every time sampled:
+        # a. >=N units were ready (possibly different each time), or
+        # b. >=N units were ready each time
+        for name in status.units:
+            if name in status.idle_units:
+                idle_since[name] = min(now, idle_since.get(name, float("inf")))
+            else:
+                idle_since[name] = float("inf")
+
+        if busy := {n for n, t in idle_since.items() if t > expected_idle_since}:
+            logger.info("Waiting for units to be idle enough: %s", busy)
+            yield False
+            continue
+
+        for app_name in apps:
+            ready_units = [
+                n for n in status.ready_units if n.startswith(f"{app_name}/")
+            ]
+            if len(ready_units) < wait_for_units:
+                logger.info(
+                    "Waiting for app %r units %s >= %s",
+                    app_name,
+                    len(status.ready_units),
+                    wait_for_units,
+                )
+                yield False
+                break
+
+            if (
+                wait_for_exact_units is not None
+                and len(ready_units) != wait_for_exact_units
+            ):
+                logger.info(
+                    "Waiting for app %r units %s == %s",
+                    app_name,
+                    len(ready_units),
+                    wait_for_exact_units,
+                )
+                yield False
+                break
+        else:
+            yield True
+
+
+def check(
+    full_status: FullStatus,
+    *,
+    apps: AbstractSet[str],
+    raise_on_error: bool,
+    raise_on_blocked: bool,
+    status: str | None,
+) -> CheckStatus | None:
+    """A single iteration of a wait_for_idle loop."""
+    for app_name in apps:
+        if not full_status.applications.get(app_name):
+            logger.info("Waiting for app %r", app_name)
+            return None
+
+    # Order of errors:
+    #
+    # Machine error (any unit of any app from apps)
+    # Agent error (-"-)
+    # Workload error (-"-)
+    # App error (any app from apps)
+    #
+    # Workload blocked (any unit of any app from apps)
+    # App blocked (any app from apps)
+    units: dict[str, UnitStatus] = {}
+
+    for app_name in apps:
+        units.update(_app_units(full_status, app_name))
+
+    for unit_name, unit in units.items():
+        if unit.machine:
+            machine = full_status.machines[unit.machine]
+            assert isinstance(machine, MachineStatus)
+            assert machine.instance_status
+            if machine.instance_status.status == "error" and raise_on_error:
+                raise JujuMachineError(
+                    f"{unit_name!r} machine {unit.machine!r} has errored: {machine.instance_status.info!r}"
+                )
```

Comment:
Suggestion: make it clear up front that we're only interested in unit.machine -- there's no elif/else coming
```suggestion
        if not unit.machine:
            continue
        machine = full_status.machines[unit.machine]
        assert isinstance(machine, MachineStatus)
        assert machine.instance_status
        if machine.instance_status.status == "error" and raise_on_error:
            raise JujuMachineError(
                f"{unit_name!r} machine {unit.machine!r} has errored: {machine.instance_status.info!r}"
            )
```